zerver/lib/events.py

```diff
@@ -349,7 +349,7 @@ def name(sub):
                         user_id in sub['subscribers']):
                     sub['subscribers'].remove(user_id)
     elif event['type'] == "presence":
-        state['presences'][event['email']] = event['presence']
+        state['presences'] = UserPresence.get_status_dict_by_user(user_profile)
```

Comment:
This will overwrite all presence info for other users!  Fix is this:

```
@@ -349,7 +349,9 @@ def apply_event(state, event, user_profile, include_subscribers):
                         user_id in sub['subscribers']):
                     sub['subscribers'].remove(user_id)
     elif event['type'] == "presence":
-        state['presences'][event['email']] = event['presence']
+        # TODO: Add user_id to presence update events / state format!
+        presence_user_profile = get_user_profile_by_email(event['email'])
+        state['presences'][event['email']] = UserPresence.get_status_dict_by_user(presence_user_prof
ile)[event['email']]
     elif event['type'] == "update_message":
         # The client will get the updated message directly
         pass
```